packages/@aws-cdk/aws-appmesh/lib/route-spec.ts

```diff
@@ -0,0 +1,241 @@
+}
+
+/**
+ * Properties specific for HTTP Based Routes
+ */
+export interface HttpRouteSpecOptions {
+  /**
+   * The criterion for determining a request match for this Route
+   *
+   * @default - matches on '/'
+   */
+  readonly match?: HttpRouteMatch;
+
+  /**
+   * List of targets that traffic is routed to when a request matches the route
+   */
+  readonly weightedTargets: WeightedTarget[];
+
+  /**
+   * The priority for the route. Routes are matched based on the specified value, where 0 is the highest priority.
+   *
+   * @default - none
+   */
+  readonly priority?: number;
+}
+
+/**
+ * Properties specific for a TCP Based Routes
+ */
+export interface TcpRouteSpecOptions {
+  /**
+   * List of targets that traffic is routed to when a request matches the route
+   */
+  readonly weightedTargets: WeightedTarget[];
+}
+
+/**
+ * All Properties for GatewayRoute Specs
+ */
+export interface RouteSpecConfig {
+  /**
+   * The spec for an http route
+   *
+   * @default - no http spec
+   */
+  readonly httpRouteSpec?: CfnRoute.HttpRouteProperty;
+
+  /**
+   * The spec for an http2 route
+   *
+   * @default - no http2 spec
+   */
+  readonly http2RouteSpec?: CfnRoute.HttpRouteProperty;
+
+  /**
+   * The spec for a grpc route
+   *
+   * @default - no grpc spec
+   */
+  readonly grpcRouteSpec?: CfnRoute.GrpcRouteProperty;
+
+  /**
+   * The spec for a tcp route
+   *
+   * @default - no tcp spec
+   */
+  readonly tcpRouteSpec?: CfnRoute.TcpRouteProperty;
+
+  /**
+   * The priority for the route. Routes are matched based on the specified value, where 0 is the highest priority
+   *
+   * @default - none
+   */
+  readonly priority?: number;
+}
+
+/**
+ * Used to generate specs with different protocols for a RouteSpec
+ */
+export abstract class RouteSpec {
+  /**
+   * Creates an HTTP Based RouteSpec
+   */
+  public static http(options: HttpRouteSpecOptions): RouteSpec {
+    return new HttpRouteSpec(options, Protocol.HTTP);
+  }
+
+  /**
+   * Creates an HTTP2 Based RouteSpec
+   *
+   */
+  public static http2(options: HttpRouteSpecOptions): RouteSpec {
+    return new HttpRouteSpec(options, Protocol.HTTP2);
+  }
+
+  /**
+   * Creates a TCP Based RouteSpec
+   */
+  public static tcp(options: TcpRouteSpecOptions): RouteSpec {
+    return new TcpRouteSpec(options);
+  }
+
+  /**
+   * List of targets that traffic is routed to when a request matches the route
+   */
+  abstract readonly weightedTargets: WeightedTarget[];
+
+  /**
+   * The priority for the route. Routes are matched based on the specified value, where 0 is the highest priority
+   */
+  abstract readonly priority?: number;
+
+  /**
+   * Called when the GatewayRouteSpec type is initialized. Can be used to enforce
+   * mutual exclusivity with future properties
+   */
+  public abstract bind(scope: cdk.Construct): RouteSpecConfig;
+
+  /**
+  * Utility method to add weighted route targets to an existing route
+  */
+  protected renderWeightedTargets() {
```

Comment:
Can you just make this a function in this file that takes the array of `WeightedTarget`s as an argument?

Reply:
Any reason why? In this comment you mentioned the `Listener` class should be responsible for rendering itself: https://github.com/aws/aws-cdk/pull/10879#discussion_r50510394

Reply:
Having a file-private function that the different `RouteSpec` classes use does not change that (it's an implementation detail).

What I don't want is an additional `protected` (which is effectively the same as `public`) method on this class. We should strive to keep the API of these classes to a minimum.